[matching.ml cleanup] structured rows

trefis opened the pull request:
Here we stop representing pattern rows as a list of patterns and instead switch to a pair of head pattern (previously the first element of the list) and the rest of the row (previously the tail of the list).

We also start using the `Pattern_head.t` type (from `Parmatch`) in places.

trefis committed:
matching: half_compiled_row: (head, patl, act) => - 12 assert false

trefis committed:
matching: move things around

trefis committed:
matching: Add intermediary types to keep track of the compilation state of pms

trefis committed:
parmatch: export Pattern_head

trefis committed:
matching: start using Pattern_head

trefis committed:
matching: formatting noise

There were some patches on the file since our last series of patch that
didn't respect the formatting.
So we're calling ocamlformat again (the same version as before,
something < 0.10) so everything rebases cleanly.

gasche commented:
(This is the first batch of changes of #9321, split because it makes sense as a unit to review on its own.)

Octachron committed:
matching: more precise argo argument

Octachron commented on lambda/matching.ml:
```diff
@@ -139,6 +139,185 @@ let all_record_args lbls =
+
+  type clause = pattern Non_empty_clause.t
+
+  val of_clause : args:(lambda * 'a) list -> General.clause -> clause
```
I am not convinced by the `args` argument: only the first element of the list is used, and there is an `assert false` lurking in the code if the list is empty. I think it would be better to have:
```ocaml
val of_clause : arg:lambda -> General.clause -> clause
```
and either [removes the assert false](https://github.com/Octachron/ocaml/commit/2b5fd38e6c1835de3d450140837a6da8cef11701) or moves it the only unclear call-site in `split_or_precompile`.

trefis replied:
We integrated a slightly modified version of your code.
Thank you!

Octachron commented on lambda/matching.ml:
```diff
@@ -139,6 +139,185 @@ let all_record_args lbls =
+
+  type clause = pattern Non_empty_clause.t
+
+  val try_no_or : Half_simple.pattern -> pattern option
```
I found myself backtracking a few time on the name `try_no_or`. Maybe the function name could reflect what it is trying to do : Something like  `promote_no_or`?

gasche replied:
The rationale was to have `assert_foo` functions and `try_foo` functions, the first which does a cast or fail and the other which returns an option. `assert_foo` does not exist here, and `try_foo` is going to go away with the polymorphic-variant version anyway, so we propose to keep this name temporarily to avoid rebase pains.

Octachron commented on lambda/matching.ml:
```diff
@@ -1688,29 +1838,36 @@ let make_variant_matching_nonconst p lab def ctx = function
+          )
+        | _ ->
+            (* I really want to assert false here. *)
+            { args; cells = [] }
```
I would prefer an assert false too.

Octachron commented on lambda/matching.ml:
```diff
@@ -1348,68 +1444,119 @@ and precompile_var args cls def k =
+    pretty_precompiled_res next nexts
+  );
+  (next, nexts)
+
```
The debug part in `split_and_precompile_*` is really noisy. I think it should be factorized, otherwise the small difference between the three variants are lost in the noise.

trefis replied:
https://github.com/ocaml/ocaml/pull/9321/commits/36961ba085d8ccf9832f24c597257471056fb01d does this I believe.

gasche replied:
@Octachron remarked that this is duplication we introduced in this PR, so it would be better fixed now. I added two more commits to do it.

Octachron commented on lambda/matching.ml:
```diff
@@ -3100,6 +3257,50 @@ let rec compile_match repr partial ctx (m : pattern_matching) =
+and compile_simplified repr partial ctx (m : Simple.clause pattern_matching) =
+  match m with
+  | { cases = []; args = [] } -> comp_exit ctx m
+  | { args = ((Lvar v as arg), str) :: argl } ->
```
The local precondition `args = ((Lvar v as arg), str) :: argl ` is new. But `compile_simplified` is only called in `do_compile_matching` which is only called by `compile_match` and `compile_half_compiled` that respect this precondition.

Octachron commented on lambda/matching.ml:
```diff
@@ -3100,6 +3257,50 @@ let rec compile_match repr partial ctx (m : pattern_matching) =
+    (m : pattern Non_empty_clause.t pattern_matching) =
+  match m with
+  | { cases = []; args = [] } -> comp_exit ctx m
+  | { args = ((Lvar v as arg), str) :: argl } ->
```
This also a new precondition, but `compile_half_compiled` is only called by `compile_flattened`, which is called by `do_for_multiple_match` on a argument list that respect this precondition too. (I have a series of commit at https://github.com/Octachron/ocaml/commits/argo_analysis that makes this point explicit).

trefis replied:
> I have a series of commit at https://github.com/Octachron/ocaml/commits/argo_analysis that makes this point explicit

Thanks, but we're not going to integrate that right now.
Let's revisit once all the other PRs have been reviewed.

Octachron commented on lambda/matching.ml:
```diff
@@ -1348,68 +1444,119 @@ and precompile_var args cls def k =
+                ~mk_action:mk_new_action ~vars:(List.map fst vars) rem_cases
+            in
+            let handler =
+              { provenance = [ [ orp (* FIXME? *) ] ];
```
What ought to be fixed here?

trefis replied:
:man_shrugging: 

The handler seems to be the same as before; and we can't remember why we added a FIXME. So we'll remove it.

Octachron commented on lambda/matching.ml:
```diff
@@ -3100,6 +3257,50 @@ let rec compile_match repr partial ctx (m : pattern_matching) =
+  match m with
+  | { cases = []; args = [] } -> comp_exit ctx m
+  | { args = ((Lvar v as arg), str) :: argl } ->
+      (* Not sure we need to do anything with arg/v *)
```
I am not sure what this comment is referring too? Is it a point above?

gasche replied:
We removed the unclear comment.

Octachron reviewed (commented):
I like the general clarification, and I am pretty confident that there is no change in behaviour (I need one or more pass of review to be sure). However, I have few questions on interfaces, names and comments:

trefis committed:
matching: remove dedicated type, use labelled arguments

trefis committed:
matching: add an assert false in divide variant

This triggers if the first column contains something other than a
variant (this is similar to the "get_key"s for other heads).

gasche committed:
matching: deduplicate the matrix pretty-printing functions

gasche committed:
matching: deduplicate the debug information printing

Octachron reviewed (approved):
The current state looks good to me: it is helpful to have the different stages of the pattern compilation reflected in the types.
